samples/RhythmGame/src/main/cpp/utils/LockFreeQueue.h

```diff
@@ -0,0 +1,142 @@
+     * We have 2 counters: readCounter and writeCounter. Each will increment until it reaches
+     * INDEX_TYPE_MAX, then wrap to zero. Unsigned integer overflow is defined behaviour in C++.
+     *
+     * Each time we need to access our data array we call mask() which gives us the index into the
+     * array. This approach avoids having a "dead item" in the buffer to distinguish between full
+     * and empty states. It also allows us to have a size() method which is easily calculated.
+     */
+
+    static constexpr bool isPowerOfTwo(uint32_t n) { return (n & (n - 1)) == 0; }
+    static_assert(isPowerOfTwo(CAPACITY), "Capacity must be a power of 2");
+    static_assert(std::is_unsigned<INDEX_TYPE>::value, "Index type must be unsigned");
+
+    /**
+     * Pop a value off the head of the queue
+     *
+     * @param val - element will be stored in this variable
+     * @return true if value was popped successfully, false if the queue is empty
+     */
+    bool pop(T &val) {
+        if (isEmpty()){
+            return false;
+        } else {
+            val = buffer[mask(readCounter++)];
+            return true;
+        }
+    }
+
+    /**
+     * Add an item to the back of the queue
+     *
+     * @param item - The item to add
+     * @return true if item was added, false if the queue was full
+     */
+    bool push(const T& item) {
+        if (isFull()){
+            return false;
+        } else {
+            buffer[mask(writeCounter++)] = item;
+            return true;
+        }
+    }
+
+    /**
+     * Get the item at the front of the queue but do not remove it
+     *
+     * @param item - item will be stored in this variable
+     * @return true if item was stored, false if the queue was empty
+     */
+    bool peek(T &item) const {
+        if (isEmpty()){
+            return false;
+        } else {
+            item = buffer[mask(readCounter)];
+            return true;
+        }
+    }
+
+    /**
+     * Get the number of items in the queue
+     *
+     * @return number of items in the queue
+     */
+    INDEX_TYPE size() const {
+
+        /**
+         * This is worth some explanation:
+         *
+         * Whilst writeCounter is greater than readCounter the result of (write - read) will always
+         * be positive. Simple.
+         *
+         * But when writeCounter is equal to INDEX_TYPE_MAX (e.g. UINT32_MAX) the next push will
+         * wrap it around to zero, the start of the buffer, making writeCounter less than
+         * readCounter so the result of (write - read) will be negative.
+         *
+         * But because we're returning an unsigned type return value will be as follows:
+         *
+         * returnValue = INDEX_TYPE_MAX - (write - read)
+         *
+         * e.g. if write is 0, read is 150 and the INDEX_TYPE is uint8_t where the max value is
+         * 255 the return value will be (255 - (0 - 150)) = 105.
+         *
+         */
+        return writeCounter - readCounter;
```

Comment:
There is a concurrency problem with code that attempts to use both counters at the "same time"—here and in 'isEmpty.' Since two consecutive reads from atomic variables is not atomic, you can end up with inconsistent results.

It's safer to store both counters in a single atomic variable of double size, and use atomic 'read-modify-update' operations on it. The drawback is potential cache line sharing between two cores, but since your main goal here is just avoiding locking and not achieving super performance, it's fine.

Reply:
Now that is clever.

Reply:
Been giving this quite a bit of thought. Whilst it's correct that "two consecutive reads from atomic variables is not atomic" I don't think it's a problem here because both `isFull()` and `isEmpty()` are used pessimistically. 

An example.

1. The queue starts full
1. Writer thread tries to push an item onto the queue by calling `push()`
1. Writer thread calls `isFull()`, which via `size()` performs the non-atomic operation by reading both `readCounter` and `writeCounter`.
1. At this exact same time the reader thread attempts to pop an item off the queue by calling `pop()`.
1. Reader thread calls `isEmpty()`, which via `size()` performs the non-atomic operation by reading both `readCounter` and `writeCounter`. 

This is a race with 2 outcomes: 

Outcome 1: Push fails

1. The writer thread finishes calling `size()` before the reader thread has updated `readCounter` in which case `isFull()` is true and the push fails. This is an acceptible outcome because the concurrent pop hasn't yet completed.

Outcome 2: Push succeeds

1. The reader thread finishes calling `size()` before the writer thread and updates `readCounter`
2. On the writer thread `isFull()` returns false (there is one space in the queue now) and the push succeeds

Am I missing something?

Reply:
What about the situation when the writer thread initiates `push', updates `writeCounter', then gets preempted. `isEmpty' now returns `true', but the data isn't in the buffer yet, so the reader receives rubbish. To fix that, `push' must first do a write, and only then 'publish' the result by increasing the counter.

A similar problem with `pop'—after it has increased `readCounter', it enables the writer thread to overwrite the data it is about to read (because `ifFull' will now be returning `false'). So again, `pop' first must read the data, and only then 'publish' this fact.

Apologies for distracting your attention on the non-problem of pair inconsistency, it just was the first thing that had struck me.

Reply:
Don, I think you can take a look at the implementation here: https://www.di.ens.fr/~guatto/papers/sbac13.pdf
What you are implementing is called "Lamport Queue," and this article provides a basic implementation using C11 atomics. You can continue to use C++11 atomic with default sequential consistency mode, it will just be even less effective than their implementation, but that should not be a problem here.

Reply:
Good point, and thank you greatly for taking the time to review. 

I thought that the postincrement operator would only increment the counter *after* the rest of the operators have been evaluated, however this is incorrect since the `a++` operator has a [precedence of 2](http://en.cppreference.com/w/cpp/language/operator_precedence) and `=` has a precedence of 16.  

So if I change the following line in `push`: 

    buffer[mask(writeCounter++)] = item;

To

    buffer[mask(writeCounter)] = item;
    ++writeCounter;

This should mean that data is only 'published' after it has actually been written, and vice versa for `pop` - does that solve the problem? 

Thanks for the Lamport Queue link. Looks pretty good, may use if this doesn't pan out.

One other question: Can you think of a way I could get a failing test for the scenario you mentioned? My current one passes every time, presumably because the chance of the writer thread being preempted immediately after `a++` is very small.

Reply:
Thanks Mikhail. Code updated accordingly. Happy for me to merge?